certmap: add OpenSSL implementation

sumit-bose opened the pull request:
The OpenSSL 1.1 API is used but there is a short macro block which
should added the needed compatibility if and older OpenSSL version is
used.

Related to https://pagure.io/SSSD/sssd/issue/3050

fidencio commented:
@sumit-bose: what's the best way to test those changes?

lslebodn commented:
I was testing patches with libcrypto and I noticed that `p11_child` was not build with libcrypto.

Do we have a ticket for this?

Yes, I know it is just partially related to this PR

sumit-bose commented:
@fidencio, building with libcrypto and running make check

@lslebodn, no, I think there is none, would you mind to open one and assign it to me?

lslebodn commented:
> @lslebodn, no, I think there is none, would you mind to open one and assign it to me?
Here you are:https://pagure.io/SSSD/sssd/issue/3489

> @fidencio, building with libcrypto and running make check
because `libsss_certmap` has bunch of unit tets :-)

lslebodn commented on src/lib/certmap/sss_certmap_attr_names.c:
```diff
@@ -30,44 +30,48 @@
+    { false, false, "2.5.4.3",                    "CN",                  "CN",                            "CN"},
+    { true,  false, "2.5.4.8",                    "ST",                  "S",                             "ST"},
+    { false, false, "2.5.4.10",                   "O",                   "O",                             "O"},
+    { false, false, "2.5.4.11",                   "OU",                  "OU"                             "OU"},
```
there is missing comma after 2nd "OU"

sumit-bose replied:
wow, good catch

lslebodn commented on src/lib/certmap/sss_cert_content_crypto.c:
```diff
@@ -19,14 +19,786 @@
+
+        list[c] = talloc_asprintf(list, "%s=%.*s", openssl_2_nss_attr_name(sn),
+                                                   (int) tmp_str_size, tmp_str);
+        BIO_reset(bio_mem);
```
There is a warning:
```
src/lib/certmap/sss_cert_content_crypto.c: In function ‘get_rdn_list’:
src/lib/certmap/sss_cert_content_crypto.c:367:9: error: value computed is not used [-Werror=unused-value]
         BIO_reset(bio_mem);
         ^~~~~~~~~
```
But if I comment out this line then test failed. We might use `(void)` for suppression.

sumit-bose replied:
I checked the return value. I think with the memory type BIO_reset() should never fail, but if it really does there might be something really wrong and it would be worth to error out.

lslebodn commented on src/lib/certmap/sss_cert_content_crypto.c:
```diff
@@ -19,14 +19,786 @@
+            ret = EINVAL;
+            goto done;
+        }
+   }
```
missing one space

sumit-bose replied:
fixed

lslebodn commented on src/lib/certmap/sss_cert_content_crypto.c:
```diff
@@ -19,14 +19,786 @@
+    int len;
+    EXTENDED_KEY_USAGE *extusage = NULL;
+
+
```
Could you remove extra space?

sumit-bose replied:
fixed

lslebodn commented on src/lib/certmap/sss_cert_content_crypto.c:
```diff
@@ -19,14 +19,786 @@
+
+    for (c = 0; c < sk_ASN1_OBJECT_num(extusage); c++) {
+        len = OBJ_obj2txt(oid_buf, sizeof(oid_buf),
+                          sk_ASN1_OBJECT_value(extusage, c) , 1);
```
could you remove space before `, 1);`

sumit-bose replied:
fixed

lslebodn commented on src/lib/certmap/sss_cert_content_crypto.c:
```diff
@@ -19,14 +19,786 @@
+        }
+
+        oids_list[c] = talloc_strndup(oids_list, oid_buf, len);
+        if(oids_list[c] == NULL) {
```
missing space after `if`

sumit-bose replied:
fixed

lslebodn commented on src/lib/certmap/sss_cert_content_crypto.c:
```diff
@@ -19,14 +19,786 @@
+    }
+
+    return ret;
+
```
unnecessary extra blank line

sumit-bose replied:
fixed

lslebodn commented on src/lib/certmap/sss_cert_content_crypto.c:
```diff
@@ -19,14 +19,786 @@
+        ret = EIO;
+        goto done;
+    }
+     if (!(X509_get_extension_flags(cert) & EXFLAG_KUSAGE)) {
```
extra space before `if`

sumit-bose replied:
fixed

lslebodn commented:
BTW do you know what is purpose of ticket https://pagure.io/SSSD/sssd/issue/2880.
Is it duplicate of https://pagure.io/SSSD/sssd/issue/3489?

sumit-bose commented:
Thank you for the review, I addressed all your comments.

About the tickets, the validation is currently done in two places, the ssh responder and p11_child. So I think https://pagure.io/SSSD/sssd/issue/2880 is justified, even if the solution will be to call p11_child from the ssh responder to validate the certificate which is the solution I currently prefer.

lslebodn commented:
@jhrozek LGTM. I would appreciate more unit test for corner cases but that can be done later.

jhrozek commented:
Adding the accepted label as per @lslebodn 's comment.

jhrozek commented:
Actually it looks like @lslebodn 's comment was meant for a different PR, removing the Accepted label again, sorry for the spam.

lslebodn commented:
master:
* a20fb9cbd5f42a6ca895aea1b84347fdfea34b89
